src/ddmd/backend/cgelem.c

```diff
@@ -2280,6 +2278,70 @@ STATIC elem * eldiv(elem *e, goal_t goal)
 
     if (OPTIMIZER)
     {
+        elem *e1 = e->E1;
+        if (tyintegral(tym) && e->Eoper == OPdiv && e2->Eoper == OPconst &&
+            e1->Eoper == OPdiv && e1->E2->Eoper == OPconst)
+        {
+            /* Replace:
+             *   (e / c1) / c2
+             * With:
+             *   e / (c1 * c2)
+             */
+            targ_llong c1 = el_tolong(e1->E2);
+            targ_llong c2 = el_tolong(e2);
+            int uns1 = tyuns(e1->E1->Ety) || tyuns(e1->E2->Ety);
+            int uns2 = tyuns(e1->Ety) || tyuns(e2->Ety);
+            if (uns1 == uns2)   // identity doesn't hold for mixed sign case
+            {
+                // The transformation will fail if c1*c2 overflows. This substitutes
+                // for a proper overflow check.
+                if (uns1 ? (c1 < 0x10000000 && c2 < 0x10000000)
+                         : (-0x1000000 < c1 && c1 < 0x1000000 && -0x1000000 < c2 && c2 < 0x1000000))
+                {
+                    e->E1 = e1->E1;
+                    e1->E1 = e1->E2;
+                    e1->E2 = e2;
+                    e->E2 = e1;
+                    e1->Eoper = OPmul;
+                    return optelem(e, GOALvalue);
+                }
+            }
+        }
+
+        if (tyintegral(tym) && e->Eoper == OPdiv && e2->Eoper == OPconst &&
+            e1->Eoper == OP64_32 &&
+            e1->E1->Eoper == OPremquo && e1->E1->E2->Eoper == OPconst)
+        {
+            /* Replace:
+             *   (64_32 (e /% c1)) / c2
```

Comment:
No. The `/` is transformed into the `/%`, because that way it can later combine `/` and `%` into one instruction. This bit here undoes that optimization.
A common problem in optimizations is one optimization undoing another. The rewrites are also done bottom up, hence the order the transformations are done.